### PR TITLE
Handle RMW_DEFAULT_DOMAIN_ID.

### PR DESCRIPTION
**Pull request by hidmic**
Precisely what the title says. Connected to https://github.com/ros2/rcl/pull/689.

### CODE REVIEW

**Comment by ivanpauno** [comment]
why is this needed?
there's similar logic in rcl to avoid passing the default domain id IIRC

**Comment by ivanpauno** [comment]
See https://github.com/ros2/rcl/pull/689/files#r443041939

File: rmw_connext_shared_cpp/src/node.cpp [hunk 1]
```diff
@@ -274,8 +274,12 @@ create_node(
     }
   }
 
+  // No custom handling of RMW_DEFAULT_DOMAIN_ID. Simply use a reasonable domain id.
   participant = dpf_->create_participant(
-    static_cast<DDS::DomainId_t>(domain_id), participant_qos, NULL,
+    static_cast<DDS::DomainId_t>(
+      domain_id != RMW_DEFAULT_DOMAIN_ID ? domain_id : 0u),
```
Comment (wjwwood) [inline rmw_connext_shared_cpp/src/node.cpp:280]: nitpick: line wrap looks unnecessary.

**Comment by hidmic** [comment]
> why is this needed? there's similar logic in rcl to avoid passing the default domain id IIRC

No there isn't. Or there shouldn't be. The bug that https://github.com/ros2/rcl/pull/689 fixed was forcing a `0` when `ROS_DOMAIN_ID=`. The point of having an invalid domain ID as default is for `rmw` implementations to decide what to do with it. We can discuss whether we want to have such sentinel or not.

**Comment by hidmic** [comment]
CI up to `rcl` and `test_rmw_implementation` against all RMW implementations:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11229/) (canary build)

**Comment by ivanpauno** [comment]
> > why is this needed? there's similar logic in rcl to avoid passing the default domain id IIRC
> 
> No there isn't. Or there shouldn't be. The bug that [ros2/rcl#689](https://github.com/ros2/rcl/pull/689) fixed was forcing a `0` when `ROS_DOMAIN_ID=`. The point of having an invalid domain ID as default is for `rmw` implementations to decide what to do with it. We can discuss whether we want to have such sentinel or not.

It has been historically used to determine in `rcl` if `ROS_DOMAIN_ID` variable has to be check or not (see https://github.com/ros2/rcl/blob/72965a0e3fa9ad0af6d0fe301a1d02edc746a1fb/rcl/src/rcl/init.c#L146-L147).
 It has never been used for letting the rmw implementation decide what to do with the default domain id (same in foxy, eloquent, dashing, etc).

Not addressing https://github.com/ros2/rcl/pull/689#discussion_r443041939 will break rmw implementations not maintained by us.

**Comment by hidmic** [comment]
> It has historically been used to determine in rcl if ROS_DOMAIN_ID variable has to be check or not

True. In `rcl`. This patch is making sure that `rmw_init()` with default options as provided by implementation does not fail.

**Comment by ivanpauno** [comment]
> If an implementation does not handle RMW_DEFAULT_DOMAIN_ID, it is broken already.

Currently, rcl never passes RMW_DEFAULT_DOMAIN_ID, so it's not already broken.

**Comment by hidmic** [comment]
> Currently, `rcl` never passes RMW_DEFAULT_DOMAIN_ID, so it's not already broken.

It is if the `rmw` can't handle it. And that's regardless of whether it's most common downstream package guards it from that situation or not.

**Pull request merged**